Create spans from transaction

dmitry-saraykin opened the pull request:
This is convenient feature to use in reactive environment where it is not possible to use thread locals across the transaction lifecycle.

elasticmachine commented:
Since this is a community submitted pull request, a Jenkins build has not been kicked off automatically. Can an Elastic organization member please verify the contents of this patch and then kick off a build manually?

felixbarny commented:
Hey Dimitry and thanks for this awesome PR!

I'll have a closer look later this day. Could you explain why you changed all `Span`properties, like `start` and `type` to be volatile? Isn't a span only accessed from a single thread?

karmi commented:
Hi @dmitry-saraykin, we have found your signature in our records, but it seems like you have signed with a different e-mail than the one used in yout Git [commit](https://github.com/elastic/apm-agent-java/pull/123.patch). Can you please add both of these e-mails into your Github profile (they can be hidden), so we can match your e-mails to your Github profile?

dmitry-saraykin commented:
Hi Felix,

In general  `Span`s and `Transaction`s can be used across multiple threads in case of multithreaded processing or reactive programming. Here is example of usage spans in multiple threads with spring webflux client
```
//Thread 1:
Span span = transaction.createSpan();
span.setType("ext");
span.name("Some micro-service call");
webClient.post().uri(...).retrieve().doOnSuccess(r-> {
    //This will happen in another thread:
    span.end();
});
```

felixbarny commented:
Jenkins test this, please.

felixbarny commented on apm-agent-core/src/main/java/co/elastic/apm/impl/transaction/Span.java:
```diff
@@ -50,21 +51,23 @@
-    private String type;
+    private volatile String type;
+    
+    private volatile Span parentSpan;
```
Add `@Nullable` annotation

felixbarny commented on apm-agent-core/src/main/java/co/elastic/apm/impl/transaction/Transaction.java:
```diff
@@ -53,8 +56,12 @@
+    /**
+     * Current span created for this transaction or null if there is no span at the moment. If there are nested spans this should contain the deepest.
+     */
+    private volatile Span currentSpan;
```
I'm not sure that approach works. For example, if you are starting two spans in parallel there could be a race condition so that one span is wrongly declared as the parent of another span when they should, in fact, be both direct children of the transaction. I'd rather remove the current span here and add `Span#createSpan` (or `Span.createChildSpan`?) as the only way of creating nested spans. I'd even consider removing `ElasticApmTracer#createSpan`, but I'm not certain about that yet.

dmitry-saraykin replied:
Agree. Also though about that.

felixbarny commented on apm-agent-core/src/main/java/co/elastic/apm/impl/transaction/Transaction.java:
```diff
@@ -77,6 +84,10 @@
      */
     private boolean noop;
+    
+    public Transaction(CoreConfiguration coreConfiguration) {
```
Remove constructor and read the span limit in `#start` via `tracer.getConfig(CoreConfiguration.class).getTransactionMaxSpans()`

felixbarny commented on apm-agent-core/src/main/java/co/elastic/apm/impl/transaction/Transaction.java:
```diff
@@ -209,6 +220,35 @@ public void setUser(String id, String email, String username) {
+     */
+    public Span createSpan() {
+        Span span;
+        span = tracer.createSpan();
```
tracer is nullable, so this needs a null check

dmitry-saraykin commented:
I've created `ElasticApm#startAsyncTransaction` method to create transactions which are not bound to the current thread and must be tracked manually.
Also added `Span#createSpan` method to create child spans.
Changed `Transaction#createSpan` to create always top level spans

felixbarny commented:
I like that approach of having a method for creating a non-thread bound transaction and span. I just think we should come up with a better name, as transactions are not inherently async when created with the `startAsyncTransaction` method. Some possibilities:

* `startNonActiveTransaction`, which would correspond to `co.elastic.apm.impl.ElasticApmTracer#activate`
* make non-active transactions the default and add a method for creating thread-bound transactions
* make non-active transactions the default and add a method for making the transaction active on the current thread, add another one to deactivate.
* make it only possible to create non-active transactions with the public API, possibly adding support for that later

felixbarny commented:
I'm working on some changes based on your PR

dmitry-saraykin commented:
- I think `startNonActiveTransaction` is not a good name because the user can think that the transaction is not going to be sent or tracked. A better name may be `startManualTransaction`. Or may be there should be some method which creates/reuses currently active transaction but tells not to `end` it automatically
- It should be possible to create transactions which are bounded to the thread as it will allow to use automatic instrumentation of all internal calls

Also found a bug with gzip. Currently agent uses zip instead of gizp and apm-server fails to parse such requests

felixbarny commented:
closing in favor of https://github.com/elastic/apm-agent-java/pull/133. @dmitry-saraykin, could you please review that PR if it covers your use case?
